### ISSUE TITLE
OS X "framework build"

### ISSUE THREAD

**Issue by jjhelmus** [conda-forge/python-feedstock#23]
There has been various discussions about building Python on OS X as a "framework build" so that it can properly access the Windows manager.  I believe this would also mean that `python` and `pythonw` are no longer the same and that a `python.app` binary is include in the package.

*  [conda-forge mailing list discussion](https://groups.google.com/forum/#!topic/conda-forge/EGS-yJzWVGA)
* [conda mailing list discussion](https://groups.google.com/a/continuum.io/d/msg/conda/g7yu_UCoHhc/_tCEYuYFCgAJ)
* [conda-forge mailing list post on python.app issue](https://groups.google.com/d/msg/conda-forge/lcxQ1SS-ljI/z5R6tTmCAAAJ)

The anaconda-recipes repository has a [python-3.5 recipe](https://github.com/ContinuumIO/anaconda-recipes/tree/master/python-3.5) which provides details on how Continuum has been building Python and may be of help.

Issue #15 is related and can likely be incorperated into this.

**Comment by jjhelmus** [comment]
`python.app` is a seperate package in the defaults channel which provides `python.app` and `pythonw` in the bin directory.  Will need to investigate if conda-forge should follow this method.

Contents of the python.app defaults package:

```
$ tar -tf python.app-1.2-py35_4.tar.bz2 
bin/.python.app-post-link.sh
bin/.python.app-pre-unlink.sh
bin/python.app
bin/pythonw
info/files
info/has_prefix
info/index.json
info/meta
info/requires
pythonapp/Contents/Info.plist
pythonapp/Contents/MacOS/python
pythonapp/Contents/PkgInfo
pythonapp/Contents/Resources/empty.lproj
pythonapp/Contents/Resources/qt_menu.nib/classes.nib
pythonapp/Contents/Resources/qt_menu.nib/info.nib
pythonapp/Contents/Resources/qt_menu.nib/keyedobjects.nib
```

**Comment by jakirkham** [comment]
Honestly, I think adding the framework will make this more confusing when building things between OS X and Linux. For that reason, I don't have much love for this. Sorry.

**Comment by jjhelmus** [comment]
* [Homebrew formular for Python 3](https://github.com/Homebrew/homebrew-core/blob/master/Formula/python3.rb)

* [Stack overflow question on the topic](http://stackoverflow.com/questions/1444543/differences-between-framework-and-non-framework-builds-of-python-on-mac-os-x)

**Comment by jjhelmus** [comment]
And [Python 3.5 MacPort Portfile](https://trac.macports.org/browser/trunk/dports/lang/python35/Portfile)

**Comment by ChrisBarker-NOAA** [comment]
Folks,

Sorry to not go look at homebrew, etc, but a few notes I want to get captured. I started to look into this at the PyCon sprints, but didn't have time to actually do anything. But I did talk about it with Ned Deily, and _think_ I understand the issues.

A) you do not need a "framework" build to access the window manager. What you need is that the host app be in an application bundle -- that is what the "python.app" thing is. -- as we know it works without a framework build.

There are (at least) two ways to get this app bundle.

1) Anaconda provides Python.app, and then a pythonw executable that is a shell script that re-directs to the app bundle. This is the easiest way to do it, but requires that pythonw be something different, and that breaks setuptools entry points, and other things like forking that I don't understand.

2) the "framework" build shipped with python.org python provides a python executable written in C, that does teh magic required to bootstrap itself into an app bundle, so that it loooks and acts like a regular old command line executable to comamnd line tools, but also looks like app bundle to teh Window manager. Thus "pythonw" is jsut a soft link to python. This has worked without issue for many hears and versions of python on OS-X.

The trick is that while you don't need a framework build to use that nifty executable, right now, the build scripts only have two options: framework, then builds the nifty wrapper executable, and "regular, that doesn't.

I don't think there is actually any advantage to a full Framework for python -- it's really designed for libraries, not complete systems, so think the best solution is to hack the build scripts to build that nifty executable with an otherwise regular old *nix build.

**Comment by jakirkham** [comment]
FWIW I'm really not bothered by solution 1. Though the idea of a C bootstrap application is a very interesting idea. Thanks for sharing that, @ChrisBarker-NOAA. Is there some way that you know of to tease that part away from the undesirable framework portions somehow?

**Comment by ChrisBarker-NOAA** [comment]
FWIW I'm really not bothered by solution 1.

I am. It's simply broken for some use cases. And it doesn't seem setuptools
has any interest in supporting pythonw.

Though the idea of a C bootstrap application is a very interesting idea.
Thanks for sharing that, @ChrisBarker-NOAA
<https://github.com/ChrisBarker-NOAA>. Is there some way that you know of
to tease that part away from the undesirable framework portions somehow?

They are in theory orthogonal, but someone needs to figure out how to build
it -- probably not that hard -- and then make a configure target for it --
maybe harder ( for me anyway)

Not my area of expertise :-(

-CHB




—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/conda-forge/python-feedstock/issues/23#issuecomment-228478621>,
or mute the thread
<https://github.com/notifications/unsubscribe/AA38YJvtUjCtGBPcv7CkJ3Dha_UxwhG0ks5qPFq2gaJpZM4I84Pv>
.

**Comment by jakirkham** [comment]
This might be of some value for 2. ( https://github.com/gldnspud/virtualenv-pythonw-osx ) Looks pretty simple.

**Comment by jakirkham** [comment]
Also, it seems there is a lot of confusion about issue ( https://github.com/conda-forge/python-feedstock/issues/15 ). I'm not sure how to clarify it better. I've just tried to rename it to avoid confusion. Feedback welcome on how to improve the title/description.

The point is the Python interpreter as we have built it here and how Continuum builds it can accidentally import stuff from the System Python if it is not present in the active conda environment. It caused me many subtle and challenging to debug issues in my pre-conda days. So, I added a patch so as to stop that behavior so it would simply fail to import stuff that was not in the conda environment. It is an incredibly simple patch (couple of lines) and we should do it for our own sanity IMHO. 😄

**Comment by jakirkham** [comment]
Also a related issue at anaconda ( https://github.com/ContinuumIO/anaconda-issues/issues/199 ).

**Comment by jjhelmus** [comment]
In PR #24 I took the MacPorts Python 3.5 setting and patches, adjusted them for use in conda so that Python is build as a framework.  This radically changes the package and may break some things downstream. Feedback is welcome.

If anyone wants to try out the package I've uploaded it to [my (jjhelmus) Anaconda cloud channel](https://anaconda.org/jjhelmus/python/files) as osx-64/python-3.5.1-1.tar.bz2.

**Comment by jjhelmus** [comment]
After testing the framework build of Python I do not think this is a workable solution.  Specifically the framework build changes where the `site-packages` directory is located from the standard unix location,`$PREFIX/lib/python3.5/site-packages/`, to a framework location, `$PREFIX/Python.framework/Versions/3.5/lib/python3.5/site-packages`. Any packages placed in the unix directory are not in the search path.

This means that all `osx` conda packages which use the unix `site-packages` directory will not work with this Python, so all the conda packages in the `defaults` and `conda-forge` channels and probably everything on Anaconda.org will be broken. Additionally, any conda package created with this Python will not work with a non-framework version due to placing the packages in the framework `site-packages` directory.  From what I can tell if we use a framework version of Python it would be necessary to create a completely new set of packages for every package which would be incompatible with any non-framework version.

It may be possible to include the unix site-packages in the search path but this is non-standard and I'm not sure how well it would work.  I do not plan to explore this option myself but if someone else wants to I would welcome it.

**Comment by jakirkham** [comment]
Right, sorry, I think I did a bad job explaining this concern before. Though having symlinks in the framework to those standard locations (assuming we go the framework route), might actually work reasonably well.

**Comment by ChrisBarker-NOAA** [comment]
Yes, I think symlinks will probably work OK -- after all, the Framework build on the Mac (python.org install) works fine with the usual tools - setuptools, pip, etc. 

However, that's only going to work if everything is done with the usual python tools -- I wouldn't be surprised if conda does a few things on its own (fix #! lines, etc.) That would break.

which is why I think the "right" solution is to build the pythonw executable in a way compatible with a regular build. And I'm pretty sure there is essentially no reason to have a Framework build at all -- it's still being done for legacy reasons, and we don't have that legacy in conda.

If I find the roundtoits, I'll see what I can hack together.

-CHB

**Comment by ChrisBarker-NOAA** [comment]
IT's been a while, and I haven't found the roundtoits, but another note:

Speaking with Ned Deily way back wen, he really wanted to get rid of the Framework Build for the python.org distribution. Maybe he will for 3.7. It's only there for legacy reasons.

It was originally done that way to better fit with the "Mac" way of doing things. But Python doesn't actually fit all that well with the "framework" model (it's not just a library), and I think the consensus is that a Framework doesn't buy us anything at all. Compatibility with the *nix builds would be more beneficial, as we've seen with conda.

**Comment by jakirkham** [comment]
Do you have an MRE or perhaps even a few MREs that demonstrate issues that the current `pythonw` strategy in `conda-forge` has not solved that the Python.org install does, @ChrisBarker-NOAA?

While I have seen this issue come up in a few places and am a macOS user, I imagine that I don’t encounter them in my own development as I don’t do nearly as much GUI development as the people running into these issues (yourself included) do. Ironically when this and other issues come up, I can maybe spend a few minutes looking at them before moving on for the same reasons (i.e. not doing GUI development). After all I’m a community contributor here in conda-forge like yourself with a day job that has some overlap with this org.

Providing clear MREs would help us both address this is issue in a more satisfactory way as we can focus on a particular problem that is causing you and others grief and at least find a workaround if not a real solution. Any ideas of good MREs to look at?

**Comment by ChrisBarker-NOAA** [comment]
uhm, what's a MRE?

But I'm guessing example is what you mean.

My use-case is wxPython, but I think we have the same issue with all Desktop GUI frame works.

If you run a wxPython app with python (in conda), you get:

```
This program needs access to the screen.
Please run with a Framework build of python, and only when you are
logged in on the main display of your Mac.
```
If you install the python.app package, that gives you a "pythonw" shell script that does work:

$ pythonw sudoku-chb.py 

That works well enough that we've gotten by this far.

But it requires "pythonw" to be in the #! line, which won't work on platforms without a pythonw.

And when you use setuptools (or raw distutils) to create a "entry point" startup script, it puts "python", not pythonw in the script.

And if you monkey patch "pythonw" in , it doesn't work.

As for ready to test examples, I have a small little wx app in the pyugrid package which is on conda-forge. Though right now I can't recall if it's set up to use setuptools entry points, as that wasn't working :-)

**Comment by jni** [comment]
@jakirkham Here is a MRE. It's not quite M but it's not big and it's R. =P (@ChrisBarker-NOAA I believe it's minimal reproducible example, but I can't find any reference to support this belief!)

The issue is that installing a console script (in this case called `framework-fail`) via [standard setuptools mechanisms](https://github.com/jni/framework-python-fail/blob/3fae70b71a2d97515d287cc2df16647959a1c1b6/setup.py#L47) fails when the script wants to launch a GUI:

```
git clone git@github.com:jni/framework-python-fail
cd framework-python-fail
conda env create -f environment.py
source activate framework  # conda <4.4
# conda activate framework  # conda >=4.4
pip install -e .
framework-fail
```

Expected result:

<img width="722" alt="screen shot 2018-05-08 at 4 58 16 pm" src="https://user-images.githubusercontent.com/492549/39743049-04f000c2-52e3-11e8-8ee5-72da17ce7fde.png">

Actual result:

```
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

**Comment by ChrisBarker-NOAA** [comment]
Thanks -- that's the issue.

But too bad about the message, because it's not that it needs a Framework build, it's that it needs a Python.app -- at least in the conda case.

and it's a setuptools bug, 'cause it doesn't use "pythonw". Last I checked the setuptools folks seemed completely uninterested in addressing it. Though maybe if someone posted a patch, they'd apply it?

However, I think the best solution is do do what's done for the Python.org builds and have the regular python executable work for GUI apps. But someone's got to take the time ot figure that out.

-CHB

**Comment by cbrnr** [comment]
I assume something has happened in the meantime, because now it seems to be possible to run GUIs, but they don't get focus until after the user tab-switched back and forth (e.g. https://github.com/napari/napari/issues/380 or https://github.com/conda-forge/mnelab-feedstock/issues/10). However, I agree with @ChrisBarker-NOAA that a solution similar to what the official Python builds uses is necessary because two different `python`/`pythonw` commands are problematic.

This shouldn't just be a conda-forge issue, so is there an existing issue and/or PR for conda/Anaconda?

**Comment by ChrisBarker-NOAA** [comment]
>  is there an existing issue and/or PR for conda/Anaconda?
I don't think so -- no PR, as no one has written up a solution.

I do know that Anaconda was aware of the issue, way back when, but never took much interest. I suspect that very few of their customers are writing Desktop GUIS on OS-X :-( 

But go post an issue, by all means.

**Comment by cbrnr** [comment]
I found the relevant issue, so no need to post a new one: https://github.com/ContinuumIO/anaconda-issues/issues/199

**Comment by hoechenberger** [comment]
Opened more than 5 years ago ☹️💩

**Comment by anton-matosov** [comment]
According to official documentation as of python 3.8 this is no longer an issue
https://docs.python.org/3.8/using/mac.html#running-scripts-with-a-gui
> With Python 3.8, you can use either python or pythonw

I have tried with conda's build of python3.8 and it opens windows without python.app hack with no issues.

**Comment by cbrnr** [comment]
The official Python build has ditched `pythonw` long ago. The docs for previous versions stated that you could use either `python` or `pythonw` (e.g. https://docs.python.org/3.1/using/mac.html#running-scripts-with-a-gui), and starting with 3.8 it seems like they finally removed `pythonw` (although AFAIR I never noticed `pythonw` on earlier versions either).

The build distributed by Anaconda doesn't include this fix. Sure you can use `python` to start a GUI, but you will run into issues such as a disabled menu bar.

**Comment by cbrnr** [comment]
Try this example (you need `qtpy` or replace it with `PyQt5`):

```
from qtpy.QtWidgets import QMainWindow, QApplication

if __name__ == '__main__':
    app = QApplication([])
    win = QMainWindow()
    win.show()
    app.exec_()
```

The click on the "python" menu in the top left of the menu bar - it won't work if you have started this example with Anaconda's `python`, but it will work if you've used `pythonw`.

With the official Python builds (including Homebrew), everything just works with `python` (there is no `pythonw` on macOS anymore).

**Comment by cbrnr** [comment]
The warning is probably also present with regular `python`, I might have turned it off as explained [here](https://stackoverflow.com/questions/18733965/annoying-message-when-opening-windows-from-python-on-os-x-10-8). But I'm digressing.

**Comment by ChrisBarker-NOAA** [comment]
ping! This recenently came up again over on the python.app feedstock:

https://github.com/conda-forge/python.app-feedstock/issues/12

Look there to see some notes from python-dev. but the TL;DR is:

The cpython developers are open to patches to "fix" this -- but someone needs to actually do the painful autoconf work :-(

If someone does want to do that, it might make sense to prototype it in the conda-forge build.
